// onesignal-api-client-core/lib/builders/notification/email.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailBuilder = void 0;
const errors_1 = require("../../errors");
//  https://documentation.onesignal.com/reference#section-email-content
class EmailBuilder {
    constructor(notification) {
        this.notification = notification;
    }
    build() {
        this.checkRequiredVariables();
        return this.notification;
    }
    //  The subject of the email.
    setEmailSubject(value) {
        this.notification.email_subject = value;
        return this;
    }
    //  The body of the email you wish to send. Typically, customers include their own HTML templates here. Must include [unsubscribe_url]
    //  in an <a> tag somewhere in the email.
    // Note: any malformed HTML content will be sent to users. Please double-check your HTML is valid.
    setEmailBody(value) {
        this.notification.email_body = value;
        return this;
    }
    //  The name the email is from. If this is not specified, this will use your default from name set in Email Setup.
    setEmailFromName(value) {
        this.notification.email_from_name = value;
        return this;
    }
    //  The email address the email is from. If this is not specified, this will use your default from email in Email Setup.
    setEmailFromAddress(value) {
        this.notification.email_from_address = value;
        return this;
    }
    //  Use a template you setup on our dashboard. You can override the template values by sending other parameters with the request.
    //  The template_id is the UUID found in the URL when viewing a template on our dashboard.
    //  Example: be4a8044-bbd6-11e4-a581-000c2940e62c
    setTemplateId(value) {
        this.notification.template_id = value;
        return this;
    }
    checkRequiredVariables() {
        if (!this.notification.email_subject) {
            throw new errors_1.OneSignalError('email_subject is required');
        }
        if (!this.notification.email_body && !this.notification.template_id) {
            throw new errors_1.OneSignalError('email_body or template_id are required');
        }
    }
}
exports.EmailBuilder = EmailBuilder;
